tests/mr/device/tracking_mr_tests.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <rmm/detail/error.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device/tracking_resource_adaptor.hpp>
+#include "mr_test.hpp"
+
+#include <gtest/gtest.h>
+
+namespace rmm {
+namespace test {
+namespace {
+using Tracking_adaptor = rmm::mr::tracking_resource_adaptor<rmm::mr::device_memory_resource>;
+TEST(TrackingTest, ThrowOnNullUpstream)
+{
+  auto construct_nullptr = []() { Tracking_adaptor mr{nullptr}; };
+  EXPECT_THROW(construct_nullptr(), rmm::logic_error);
+}
+
+TEST(TrackingTest, Empty)
+{
+  Tracking_adaptor mr{rmm::mr::get_current_device_resource()};
+  EXPECT_EQ(mr.get_num_outstanding_allocations(), 0);
+}
+
+TEST(TrackingTest, AllFreed)
+{
+  Tracking_adaptor mr{rmm::mr::get_current_device_resource()};
+  std::vector<void *> allocations;
+  for (int i = 0; i < 10; ++i) {
+    allocations.push_back(mr.allocate(10_MiB));
+  }
+  for (auto p : allocations) {
+    mr.deallocate(p, 10_MiB);
+  }
+  EXPECT_EQ(mr.get_num_outstanding_allocations(), 0);
+}
+
+TEST(TrackingTest, AllocationsLeft)
+{
+  Tracking_adaptor mr{rmm::mr::get_current_device_resource()};
+  std::vector<void *> allocations;
+  for (int i = 0; i < 10; ++i) {
+    allocations.push_back(mr.allocate(10_MiB));
+  }
+  for (int i = 0; i < 10; i += 2) {
+    mr.deallocate(allocations[i], 10_MiB);
+  }
+  EXPECT_EQ(mr.get_num_outstanding_allocations(), 5);
+  mr.print_outstanding_allocations();
+}
+
```

Comment:
Test the backtrace functionality?

Reply:
I didn't know of a good way to verify the backtrace. Any suggestions here?

Reply:
I think just verify that there is some text. Don't verify that the text is correct?

Reply:
instead of `print_outstanding_allocations`, should I make the function return a `std::string` for you to do with as you please and then use that for the test?

Reply:
I think there should be two functions: `log_outstanding_allocations`, and `get_outstanding_allocations` which returns the map of pointers to size+stack trace pairs. That way it can be used programmatically as well as logged.